### PR TITLE
Update to `wgpu` 0.17.0

### PR DESCRIPTION
**Pull request by Aaron1011**
This required bumping wasm-bindgen to 0.2.87

### CODE REVIEW

**Comment by Chickenkeeper** [comment]
Would it be possible to remove the minor part of the version number, i.e. `"0.17"`, to match most of the other dependencies? The minor versions are only used for non-breaking updates, which get missed unless egui can be updated every time wgpu does

**Comment by Mingun** [comment]
Cargo specification [already](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field) [means](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility) `^0.17.0`, which is shorthand for `>=0.17.0,<0.18.0`

**Comment by Wumpf** [comment]
This actually breaks the web build because of wgpu's Send/Sync issues. @Aaron1011 do you have time to look into that?

**Comment by Chickenkeeper** [comment]
> Cargo specification [already](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field) [means](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility) `^0.17.0`, which is shorthand for `>=0.17.0,<0.18.0`

Oh that's good to know, I was under the impression that it was equal to `=0.17.0` instead. Thanks for the links, I'll read up on it properly

**Comment by Aaron1011** [comment]
This is tricky - the `type_map::concurrent::TypeMap` instance that egui exposes requires `Send + Sync` tpyes, which means that you can no longer store wgpu types in it on wasm.

I think our two options are:
* Use a non-concurrent `TypeMap` in the exposed field. This will complicate multithreaded non-wasm use cases, as references to that `TypeMap` Wil lno longer be able to be shared between threads.
* Provide two `TypeMap` instances - one non-concurrent, and one `type_map::concurrent::TypeMap`. This will require adding a parameter to some user closures, which might break lots of user code. However, it will provide more flexibility, and make it easier for users to write code targeting wasm.

**Comment by lucasmerlin** [comment]
I solved the typemap problem in my project by choosing the concurrent version based on the target_arch: 
```rs
#[cfg(not(target_arch = "wasm32"))]
use type_map::concurrent::TypeMap;
#[cfg(target_arch = "wasm32")]
use type_map::TypeMap;
```
Maybe this would work for egui as well. Since most wasm code is single threaded anyways, I don't think it'd be necessary to have the concurrent typemap on wasm.

**Comment by Aaron1011** [comment]
@lucasmerlin: My concern is that doing that would make it unnecessarily difficult for downstream crates to compile against wasm - using `TypeMap` in a function/closure signature (e.g. for `PrepareCallback`)  would require a similar `#[cfg]` in downstream crates.

**Comment by Wumpf** [comment]
Huh, tricky. Agreed that the "infectious" `cfg` isn't great either.
Tbh I'm quite unhappy with the interface having this typemap in there in the first place. Maybe we should bite the bullet and come up with something different 🤔
Need to update wgpu over on https://github.com/rerun-io/rerun/ soon as well, so I might be spending some time on this as well in the next couple of days, but I'd be very happy for more input.
Pondering if we should go with the stop-gap-solution of enabling `fragile-send-sync-non-atomic-wasm` for the time being (should then ofc referencing new bug ticket explaining the [situation](https://github.com/gfx-rs/wgpu/blob/trunk/CHANGELOG.md#wgpu-types-now-send-sync-on-wasm))

**Comment by Aaron1011** [comment]
I believe that https://github.com/gfx-rs/wgpu/pull/3626 will unblock some lifetime constraint relaxtions (see https://github.com/gfx-rs/wgpu/issues/1453). This might remove the need for egui to provide a `TypeMap` at all.

**Comment by Wumpf** [comment]
That might very well be, but https://github.com/gfx-rs/wgpu/pull/3626 has been in the works for _very_ long and may not land all that soon. Also will definitely be a major release of wgpu, not a patch; so we're still stack with pre-arcanization

**Comment by Speak2Erase** [comment]
While getting rid of the typemap is nice, even after https://github.com/gfx-rs/wgpu/pull/3626 lands it could take a while for them to fix https://github.com/gfx-rs/wgpu/issues/1453

I'm really hoping that comes sooner rather than later though. The typemap setup is very annoying to work with when doing custom rendering.

**Comment by cwfitzgerald** [comment]
If everything goes to plan, expect to see both changes in 0.18. No hard promises though.

**Comment by fornwall** [comment]
As said above, perhaps `features = ["fragile-send-sync-non-atomic-wasm"]` is the best way forward to unblock the update for now?

We are no worse off than with wgpu 0.16.0, and a proper fix can hopefully be included in a future wgpu 0.18.0/0.19.0 update.

**Comment by emilk** [comment]
Sounds good to me @fornwall

**Comment by fornwall** [comment]
@Aaron1011 Is setting `features = ["fragile-send-sync-non-atomic-wasm"]` in this PR something you have time to do?

File: crates/eframe/Cargo.toml [hunk 1]
```diff
@@ -140,7 +140,7 @@ winapi = "0.3.9"
 bytemuck = "1.7"
 js-sys = "0.3"
 percent-encoding = "2.1"
-wasm-bindgen = "0.2.86"
+wasm-bindgen = "0.2.87"
```
Comment (emilk) [inline crates/eframe/Cargo.toml:143]: You should also update this in `.github/workflows/rust.yml` and `scripts/setup_web.sh`

Comment (emilk) [reply crates/eframe/Cargo.toml:143]: I'll do it in a separate PR to minimize the diff of this PR:
* https://github.com/emilk/egui/pull/3237

**Comment by emilk** [comment]
Tested natively on Mac, as well as on Web.

**Comment by Aaron1011** [comment]
Sorry for the delay in getting to this - thank you for taking care of this!

**Comment by daxpedda** [comment]
> As said above, perhaps `features = ["fragile-send-sync-non-atomic-wasm"]` is the best way forward to unblock the update for now?
> 
> We are no worse off than with wgpu 0.16.0, and a proper fix can hopefully be included in a future wgpu 0.18.0/0.19.0 update.

FWIW `egui` used to work fine with Wasm Atomics, after this PR it won't compile with it.

In any case, great work here!

**Pull request merged**